### PR TITLE
First try at periodic db export

### PR DESCRIPTION
**Pull request by maufl**
Hi, first a disclaimer, I don't regularly program in Java and even less so for Android.
I try to implement an automatic export/backup of the database. The idea is to be able to specify where the database should be written every nth hour or so. Then, a sync tool like Syncthing picks it up and transfers it elsewhere, I import it into Influx DB or whatever and make nice graphs using Grafana or something similar. I think it is related to #553.

I started to add preferences, added an broadcast receiver, schedule intents using the AlarmManager and extended the export to an arbitrary URI so that the DB can be exported to any location a DocumentProvider supports.

There is still one big problem and probably a lot of room for improvement of the code style.
I'm unable to display the location of the exported file properly, I only receive a display name from the provider but no path. Furthermore, if I use a CREATE_DOCUMENT intent I can't pick an already existing file, and if I use a OPEN_DOCUMENT intent I can't create a new file :(
This could be solver by using a traditional file picker, however I don't know how or more precisely, it seems this would require a third party library and I don't know which I would use.
However, using the new Storage Access Framework seems to be the better option anyway, as this would for example allow to write the file directly to Google Drive, Dropbox or Nextcloud/Owncloud.

I would be very happy if somebody could provide some help so that I can get this PR into a state where it can be merged. Please let me know what I should change and improve. (Some things I already know, like setting the summary of the new preferences, scheduling the intent again after reboot).

Todo
* [x] Reschedule export broadcast after boot
* [x] What happens to the DB management activity?
* [x] Test export failed notification

### CODE REVIEW

**Comment by cpfeiffer** [comment]
Wow, great job, thanks a lot, this looks very good already. :+1: 

I'll try to answer your questions and add some minor review comments to your commit.

1. Wouldn't it be sufficient choose a directory and decide for the filename yourself? Then the CREATE vs OPEN issue would not matter. See https://stackoverflow.com/questions/27898676/simple-android-directory-picker-how for an example directory picker.

2. For directly uploading the file, I'm not sure if we would need the internet permission, even if it would be handled by another app?

**Comment by maufl** [comment]
Thanks for your help! I'll work on your reviews tomorrow.
Regarding your questions:
1. I tried that first but was unable to get a DISPLAY_NAME or something similar to show to the user. In fact this also bugs me about the file picking. I can't find out how to show the storage provider and full path to the user, so that she/he knows where the database will be :( Furthermore, I read that the application acquires permissions on the file by opening it through the intent. I'm not sure if opening a directory would grant permission to write a file there but I'll look into it.
2. From the documentation I would assume that the content provider needs the required permissions to upload the file and not us. I'll test it.

**Comment by cpfeiffer** [comment]
I think this problem would go away with requesting the WRITE_EXTERNAL_STORAGE permission.

**Comment by maufl** [comment]
But only if the file will actually be written to the external storage, right? In this case ACTION_OPEN_DOCUMENT_TREE could also return a folder in Google Drive or Dropbox which is not stored on the external storage. At least that's what understood when reading the DocumentProvider documentation.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java [hunk 1]
```diff
@@ -261,18 +269,48 @@ public boolean onPreferenceClick(Preference preference) {
...
-            public boolean onPreferenceChange(Preference preference, Object newVal) {
-                // reset city id and force a new lookup
-                GBApplication.getPrefs().getPreferences().edit().putString("weather_cityid",null).apply();
-                preference.setSummary(newVal.toString());
-                Intent intent = new Intent("GB_UPDATE_WEATHER");
-                intent.setPackage(BuildConfig.APPLICATION_ID);
-                sendBroadcast(intent);
+            public boolean onPreferenceChange(Preference preference, Object auto_export_period) {
+                preference.setSummary(auto_export_period.toString());
+                boolean auto_export_enabled = GBApplication.getPrefs().getBoolean("auto_export_enabled", false);
```
[the remark this answers is not in the document]
Comment (maufl) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: Should they be added as constants to GBPrefs or as resources to res/values.xml? The last one seems to be more the Android way but the first one contains more recent changes.

**Comment by maufl** [comment]
Well, this storage access framework drives me up the wall. It seems that even though the documents provided by the provider are expected to live in a [(virtual) document tree](https://developer.android.com/guide/topics/providers/document-provider.html#overview) there is no way of building a full path to the document that you can display to a user, because (the document provide only defines a forward mapping)[https://developer.android.com/reference/android/support/v4/provider/DocumentFile.html#getParentFile()] and my test showed that `getParentFile()` returns `null` :(

So to me there are two possibilities:
1. Do not use the storage access framework but a traditional file picker. We then need the WRITE_EXTERNAL_STORAGE permission and can only write to the "sd card" but we are able to show the user where the file will be written.
2. Use the storage access framework. We will not be able to show the user where the file is written (but we could start an intent to open it) after she/he picked it. On the plus side, the user could pick a file on GoogleDrive/Dropbox/Nextcloud, we would write to it like a regular file but it would automatically be uploaded.

If someone has more experience using the storage access framework please chime in, it's driving me crazy.

Another question that came up is, what should happen to the database management activity? Should it also use the export location picked for the periodic export? Should it be removed completely?

**Comment by cpfeiffer** [comment]
Interesting findings! I have no experience with the storage access framework yet, so I'm not much of a help without trying it myself (can't right now).

From your comments, the "documents" appear to be opaque to the client, which makes sense in a way (the storage being an implementation detail of the provider).

I don't synchronise data to external storage, so I can't comment on the usability of this approach. For flexibility we could offer both approaches (file based and provider-based, if the file based approach would not be sufficient.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cpfeiffer) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: GBPrefs should be sufficient. Resources are mostly for localizable things, AFAIU.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/FileUtils.java [hunk 2]
```diff
@@ -111,15 +117,23 @@ public static void copyURItoFile(Context ctx, Uri uri, File destFile) throws IOE
...
     public static void copyFileToURI(Context context, File src, Uri dst) throws IOException {
         OutputStream out = context.getContentResolver().openOutputStream(dst);
         if (out == null) {
             throw new IOException("Unable to open output stream for " + dst.toString());
         }
         try (OutputStream bufOut = new BufferedOutputStream(out)) {
             copyFileToStream(src, bufOut);
-            bufOut.close();
         }
+        out.close();
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/FileUtils.java]: This is also not necessary -- closing the `BufferedOutputStream` using the try-with-resources statement will automatically close the wrapped `out` stream as well.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (maufl) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: I just noticed that there would be two places that define the preference key then, the GBPrefs and the preferences.xml. That doesn't feel right to me.

[the remark this answers is not in the document]
Comment (maufl) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: On the other hand, string resources might accidentally be translated ... So I'll go with the GBPrefs constants.

**Comment by maufl** [comment]
I'm quite confident that it's now in a useful state. I gave up on finding a proper solution for displaying the file path and found one on Stackoverflow that is a bit hackisch but works. I accidentally tested the export failed notification and it works an brings you to the settings. I also fixed a bug where the export would stop working after reboot due to missing permissions. And I'm reenabling the export after a reboot now.

**Comment by cpfeiffer** [comment]
:+1:  I'll have a look at your code and most probably merge it. 

Regarding the preferences constants, you were actually right. We do have a `values.xml` with this comment right on top:
```<!-- Note: these are not translated! They are used as preference keys -->```
OTOH, `GBPrefs.FOO` looks nicer than `getContext().getString(R.p_foo)`

Regarding the db management activity: I want to keep the option to export and import manually, especially for debugging. This one also exports the preferences, IIRC. 

Concerning the location of the files: while debugging things, I probably wouldn't want to overwrite files in my synchronization directory, so we should better keep these separate.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java [hunk 3]
```diff
@@ -261,18 +274,59 @@ public boolean onPreferenceClick(Preference preference) {
...
             @Override
-            public boolean onPreferenceChange(Preference preference, Object newVal) {
-                // reset city id and force a new lookup
-                GBApplication.getPrefs().getPreferences().edit().putString("weather_cityid",null).apply();
-                preference.setSummary(newVal.toString());
-                Intent intent = new Intent("GB_UPDATE_WEATHER");
-                intent.setPackage(BuildConfig.APPLICATION_ID);
-                sendBroadcast(intent);
+            public boolean onPreferenceChange(Preference preference, Object autoExportInterval) {
+                String summary = String.format(
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: JFYI (no need to change this): getString() is overloaded with a var-args Object parameters, so you can directly add arguments to the String without the need to use String.format(). The placeholder format is a bit different though (e.g. `%1$s` for the first string argument).

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java [hunk 4]
```diff
@@ -301,6 +355,109 @@ public boolean onPreferenceChange(Preference preference, Object newVal) {
...
+            return cursor.getString(cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME));
+        }
+        return "";
+    }
+
+    /*
+    As seen on stackoverflow https://stackoverflow.com/a/36714242/1207186
+    Try to find the file path of a document uri
+     */
+    private String getFilePath(Context context, Uri uri) throws URISyntaxException {
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: We can make that a static method in class `AndroidUtils`.

Comment (maufl) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/SettingsActivity.java]: Done! :)

**Comment by maufl** [comment]
I just noticed that using GRANT_PERSISTABLE_URI_PERMISSION allows you to select the same file again and overwrite it! This was one of the last paper cuts for me.

**Comment by cpfeiffer** [comment]
Awesome! :heart: :tada:

**Pull request merged**